### PR TITLE
Add users without tax number to webapp

### PR DESCRIPTION
**Pull request by JulianRoesner**
# Short Description
We want to allow users in the webapp to submit taxes without a tax number. For this we need to request the list of tax offices from Erica. Additionally, we need to ask users if they have a tax number and ask them then what tax office they have.

# Changes
- Add a StepChooser for the lotse flow
- Add a new SteuerlotseStep for the StepSteuernummer
- Add `request_tax_office` to erica_client
- Add a new `form_steuernummer.html`
- Add javascript utility functions to show or hide fields
- Add Javascript to move through the different states of the steuernummer page
- Remove jquery from person_b and familienstand html
- Adapt acceptance tests

# Feedback
- Is the state management what you would have expected?
- Anything else you can spot?

### CODE REVIEW

File: webapp/app/templates/lotse/form_steuernummer.html [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            let bufaNrDropdown = document.getElementById("bufa_nr");
+            var selectedState = stateDropdown.options[stateDropdown.selectedIndex].value.toLowerCase();
+            var selectedTaxOffices = taxOffices.find(taxOffice => taxOffice.state_abbreviation === selectedState)
+            let oldBufaNrValue = bufaNrDropdown.value
+            deleteSelectOptions(bufaNrDropdown, defaultOptionText="{{ _('form.select_input.default_selection') }}");
+
+            if(selectedTaxOffices){
+                selectedTaxOffices = selectedTaxOffices.tax_offices;
+                selectedTaxOffices && selectedTaxOffices.sort(function (a, b) {
+                    return ('' + a.name).localeCompare(b.name);
```
Comment (frederike-ramin) [inline webapp/app/templates/lotse/form_steuernummer.html:68]: The `'' +` probably has a reason here? So a comment could help explain that?

File: webapp/app/templates/lotse/form_steuernummer.html [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        function updateTaxOfficeList() {
+            let stateDropdown = document.getElementById("bundesland");
+            let bufaNrDropdown = document.getElementById("bufa_nr");
+            var selectedState = stateDropdown.options[stateDropdown.selectedIndex].value.toLowerCase();
+            var selectedTaxOffices = taxOffices.find(taxOffice => taxOffice.state_abbreviation === selectedState)
+            let oldBufaNrValue = bufaNrDropdown.value
+            deleteSelectOptions(bufaNrDropdown, defaultOptionText="{{ _('form.select_input.default_selection') }}");
+
+            if(selectedTaxOffices){
+                selectedTaxOffices = selectedTaxOffices.tax_offices;
```
Comment (frederike-ramin) [inline webapp/app/templates/lotse/form_steuernummer.html:66]: I think I'd be more plausible to me to do `selectedTaxOffices.tax_offices` right in line 48 instead of line 53. (This way we can avoid calling a dict of states and tax offices also just "selectedTaxOffices".

File: webapp/app/templates/lotse/form_steuernummer.html [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                }
+                bufaNrDropdown.value = oldBufaNrValue
+            }
+        }
+
+        function changeTaxNumberExists(){
+            let taxNumberExistsYesField = document.querySelector('input[type=radio][name=steuernummer_exists][value="yes"]');
+            let taxNumberExistsNoField = document.querySelector('input[type=radio][name=steuernummer_exists][value="no"]');
+            if (taxNumberExistsYesField.checked && !taxNumberExistsYesField.wasCheckedBefore){
+                console.log("dudu")
```
Comment (frederike-ramin) [inline webapp/app/templates/lotse/form_steuernummer.html:82]: There it is again^^

File: webapp/app/templates/lotse/form_steuernummer.html [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+        function runThroughAllStates(){
+            changeTaxNumberExists();
+            changeBundesland();
+            changeTaxOffice();
+        }
+        document.addEventListener('DOMContentLoaded', () => {
+            // Register event handlers
+            document.querySelectorAll('input[type=radio][name=steuernummer_exists]').forEach((field)=>{
+                field.addEventListener("click",() => {changeTaxNumberExists(); runThroughAllStates(); handleState();});
```
Comment (frederike-ramin) [inline webapp/app/templates/lotse/form_steuernummer.html]: Why do you use `click` here instead of `change`? I think we could get rid of the `wasCheckedBefore` attribute if we change this?

File: webapp/app/templates/lotse/form_steuernummer.html [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+            if (bundeslandField.value && taxNumberFormState === TAX_NUMBER_FORM_STATES.StateSelectionTaxNumberExists){
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.TaxNumberInput;
+            }else if(bundeslandField.value
+                     && (taxNumberFormState === TAX_NUMBER_FORM_STATES.StateSelectionNoTaxNumberExists
+                            || taxNumberFormState === TAX_NUMBER_FORM_STATES.TaxOfficeSelection)){
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.TaxOfficeSelection;
+                updateTaxOfficeList();
+            }else if(bundeslandField.value && taxNumberFormState === TAX_NUMBER_FORM_STATES.RequestNewTaxNumber){
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.TaxOfficeSelection;
+                updateTaxOfficeList();
```
Comment (frederike-ramin) [inline webapp/app/templates/lotse/form_steuernummer.html:104]: Why aren't thos two cases the same?

File: webapp/app/templates/lotse/form_steuernummer.html [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+                    deleteInput(stateDropdown);
+                    hideRequiredInput(stateDropdown);
+                    deleteInput(bufaNrDropdown);
+                    hideRequiredInput(bufaNrDropdown);
+                    deleteInput(taxNumberInput);
+                    hideRequiredInput(taxNumberInput);
+                    deleteInput(requestNewTaxNumberCheckbox);
+                    hideRequiredInput(requestNewTaxNumberCheckbox);
+                    break;
+                case TAX_NUMBER_FORM_STATES.StateSelectionTaxNumberExists:
```
Comment (frederike-ramin) [inline webapp/app/templates/lotse/form_steuernummer.html:134]: Okay, this might be a provocative question, but why do we need the states at all? First I thought I liked it but then I saw this big switch-case and now I'm wondering: Why don't we just call these functions on e.g. dropdown change, so in the event handlers? I'm sure there is a great reason and I just don't see it yet.

File: webapp/tests/model/test_form_data.py [hunk 7]
```diff
@@ -154,24 +166,56 @@ def setUp(self) -> None:
...
 
+    def test_if_steuernummer_and_steuernummer_then_raise_missing_error(self):
+        invalid_tax_nr_data = {
+            'steuernummer_exists': True,
+            'bundesland': 'BY',
+        }
+        with self.assertRaises(ValidationError) as validation_error:
+            MandatoryFormData.parse_obj({**self.valid_data_person_a, **self.valid_data_person_b, **self.married_familienstand, **invalid_tax_nr_data} )
+        self.assertIsInstance(validation_error.exception.raw_errors[0].exc, MissingError)
+        self.assertEqual('steuernummer', validation_error.exception.raw_errors[0]._loc)
```
Comment (frederike-ramin) [inline webapp/tests/model/test_form_data.py:181]: Probably `test_if_steuernummer_exists_and_no_steuernummer_given_then_raise_missing_error`

**Review by frederike-ramin — COMMENTED** [review]
Looks good so far! 
I just have a few questions and smaller suggestions.

Additionally, with Baden-Württemberg, a strange thing is happening: If I select "no tax number" then "Baden-Württemberg"  and then directly press send, the tax office gets filled with "Offenburg Außenstelle Achern". Is there maybe a wrong default set or something? (it doesn't matter what I selected before)

File: acceptance_tests/cypress/integration/lotse_flow.spec.js [hunk 8]
```diff
@@ -287,6 +288,55 @@ context('Acceptance tests', () => {
...
+                cy.get('label[for=confirm_complete_correct]').first().click()
+                cy.get(submitBtnSelector).click()
+                cy.get('label[for=confirm_data_privacy]').first().click()
+                cy.get('label[for=confirm_terms_of_service]').first().click()
+                cy.get(submitBtnSelector).click()
+
+                // Verify success.
+                cy.get('body').contains('Ihre Informationen wurden erfolgreich verschickt.')
+                // Get PDF - can't click on it as it opens a new window, so we request it directly
+                cy.request('/download_pdf/print.pdf').its('body').should('not.be.empty')
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply acceptance_tests/cypress/integration/lotse_flow.spec.js:335]: Great idea

File: webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+    def _set_bufa_choices(self, tax_offices):
+        choices = []
+        for county in tax_offices:
+            choices += [(tax_office.get('bufa_nr'), tax_office.get('name')) for tax_office in county.get('tax_offices')]
+        self.InputForm.bufa_nr.kwargs['choices'] = choices
+
+    def _pre_handle(self):
+        tax_offices = request_tax_offices()
+        self._set_bufa_choices(tax_offices)
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py:147]: Nope. We have to set them for the validation in WTForms. If we do not set them, WTForms won't except any input because those are not valid choices

[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py:147]: Maybe I should add a comment for that?

File: webapp/app/templates/lotse/form_familienstand.html [hunk 10]
```diff
@@ -46,6 +47,7 @@
             const inputted_date = convertStringToDate(document.querySelector(selector + '_1').value,
                 document.querySelector(selector + '_2').value,
                 document.querySelector(selector + '_3').value)
+            console.log("dudu");
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/templates/lotse/form_familienstand.html]: That comment was worth it

File: webapp/app/templates/lotse/form_steuernummer.html (hunk 6 above)
Comment (JulianRoesner) [reply webapp/app/templates/lotse/form_steuernummer.html:134]: Imo there are two advantages of the states:

1. They decouple the logic and the visual changes. This makes it very easy to change how something should look like.

2. We can just go through all the different functions to check the current state and the change everything and delete data in every state. BTW: I could imagine that the flickering without the hidden class in Familienstand might happen because we change the state a few times.
This also decouples the states from each other

File: webapp/app/templates/lotse/form_steuernummer.html (hunk 1 above)
Comment (JulianRoesner) [reply webapp/app/templates/lotse/form_steuernummer.html:68]: It is to make it a String, but I think we could make the conversion explicit and then we would need no comment

File: webapp/app/templates/lotse/form_steuernummer.html (hunk 2 above)
Comment (JulianRoesner) [reply webapp/app/templates/lotse/form_steuernummer.html:66]: I See your point. I just wanted to make sure that we have found something. So directly in line 48 wouldn't work.
Just to clarify: It is not a dict of states and tax offices but an object for a specific selected state with state name, abbreviation and the tax offices.
That said, I get your point. So maybe we could just name it selectedState in line 48 and the selectedTaxOffices in line 53?

File: webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py:147]: Yes, a comment is probably a good idea. And of course, that makes total sense. 👌

**Comment by frederike-ramin** [comment]
And I forgot to write down one thing I already told you: When we select "existing tax number" and send with valid data, in the summary, there still is a line for "Neue StNr beantragen: Nein", we probably don't want that.

File: webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                       'input_req_err_msg': _l('form.lotse.field_bundesland_required')}
+        )
+        bufa_nr = SteuerlotseSelectField(
+            label=_l('form.lotse.bufa_nr'),
+            choices=[
+                ('', '---'),
+            ],
+            validators=[require_bufanr_input],
+            render_kw={'data_label': _l('form.lotse.bufa_nr.data_label')}
+        )
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py:127]: So good 😊

File: webapp/app/templates/lotse/form_steuernummer.html (hunk 5 above)
Comment (JulianRoesner) [reply webapp/app/templates/lotse/form_steuernummer.html:104]: You are totally correct. They should be the same.

File: webapp/app/templates/lotse/form_familienstand.html [hunk 12]
```diff
@@ -6,35 +6,36 @@
     <div id="familienstand_field" class="mb-3">
         {{ macros.render_field(form.familienstand, 10) }}
     </div>
-    <div id="familienstand_date_field" class="form-row-center hidden">
+    <div id="familienstand_date_field" class="form-row-center">
         {{ macros.render_field(form.familienstand_date) }}
     </div>
-    <div id="familienstand_married_lived_separated_field" class="form-row-center hidden">
+    <div id="familienstand_married_lived_separated_field" class="form-row-center">
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/templates/lotse/form_familienstand.html:12]: I was worried that something like that might happen. The only reason was to get rid of the jquery. I'll have a quick look

File: webapp/app/templates/lotse/form_steuernummer.html [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+        {{ macros.render_field(form.request_new_tax_number) }}
+    </div>
+
+{% endblock %}
+
+{% block optional_js %}
+    {{ super() }}
+    <script src="{{ url_for('static', filename='js/field_utils.js') }}" type=text/javascript></script>
+    <script>
+        let taxOffices = JSON.parse('{{ render_info.additional_info['tax_offices']|tojson }}');
```
Comment (nfelger) [inline webapp/app/templates/lotse/form_steuernummer.html:33]: The value of `taxOffices` does not change, so it should be `const` not `let`. Applies to the rest of the variable declarations in this file. Basically, variable declarations should be `const` _unless_ you need to write `varname = newValue` anywhere.

File: webapp/app/templates/lotse/form_steuernummer.html [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+            handleState();
+        });
+
+        function updateTaxOfficeList() {
+            let stateDropdown = document.getElementById("bundesland");
+            let bufaNrDropdown = document.getElementById("bufa_nr");
+            var selectedState = stateDropdown.options[stateDropdown.selectedIndex].value.toLowerCase();
+            var selectedTaxOffices = taxOffices.find(taxOffice => taxOffice.state_abbreviation === selectedState)
+            let oldBufaNrValue = bufaNrDropdown.value
+            deleteSelectOptions(bufaNrDropdown, defaultOptionText="{{ _('form.select_input.default_selection') }}");
```
Comment (nfelger) [inline webapp/app/templates/lotse/form_steuernummer.html:63]: Why are some of these declared with `var` and some with `let`?

I'd prefer if we didn't mix "new"-style `let`/`const` with old-style `var`. Actually, I'd prefer if we didn't `var` at all unless there's a very very very very good reason for it.

File: webapp/app/templates/lotse/form_steuernummer.html [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        }
+
+        function changeTaxNumberExists(){
+            let taxNumberExistsYesField = document.querySelector('input[type=radio][name=steuernummer_exists][value="yes"]');
+            let taxNumberExistsNoField = document.querySelector('input[type=radio][name=steuernummer_exists][value="no"]');
+            if (taxNumberExistsYesField.checked && !taxNumberExistsYesField.wasCheckedBefore){
+                console.log("dudu")
+                taxNumberFormState = TAX_NUMBER_FORM_STATES.StateSelectionTaxNumberExists;
+                taxNumberExistsYesField.wasCheckedBefore = true;
+                taxNumberExistsNoField.wasCheckedBefore = false;
```
Comment (nfelger) [inline webapp/app/templates/lotse/form_steuernummer.html:85]: I like using the DOM for storing UI state 👍

File: webapp/app/templates/lotse/form_steuernummer.html (hunk 6 above)
Comment (nfelger) [reply webapp/app/templates/lotse/form_steuernummer.html:134]: I do really like the explicit state modelling, but I also think the boundaries don't sit in the ideal places yet. The giveaway for me is less the switch statement here (here we have a single, clearly-delineated code block for each sate) but more the many if-branches in `changeTaxNumberExists`/`changeBundesland`/`changeTaxOffice`, which I've found pretty hard to follow.

I'm not sure what a better approach would be, but since we're moving towards React, there's probably not too much point in working this out at this point…

File: webapp/app/templates/lotse/form_steuernummer.html (hunk 2 above)
Comment (nfelger) [reply webapp/app/templates/lotse/form_steuernummer.html:66]: A bit left-field for this PR, but we _could_ also change the structure of the JSON to be an object that has state abbreviations as keys and a list of tax offices as the value for each of them…

File: webapp/tests/forms/flows/test_lotse_step_chooser.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    @pytest.fixture(autouse=True)
+    def attach_fixtures(self, test_request_context):
+        self.req = test_request_context
+
+        self.testing_steps = [
+            StepSteuernummer,
+        ]
+        self.endpoint_correct = "lotse"
+
+    def test_set_attributes_correctly(self):
```
Comment (nfelger) [inline webapp/tests/forms/flows/test_lotse_step_chooser.py:18]: I would question whether we need this test at all. It mainly tests functionality of the base class. Just throwing it out there for you to consider.

File: webapp/app/static/js/field_utils.js [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    input.value = '';
+  }
+}
+function hideInput(input) {
+  input.style.setProperty("display", "none", "important");
+  if (input.associatedContent) {
+    input.associatedContent.forEach((associatedElem) =>{
+      associatedElem.oldDisplayStyle = "inline-block";
+      associatedElem.style.setProperty("display", "none", "important");
+
```
Comment (nfelger) [inline webapp/app/static/js/field_utils.js:37]: Curious: did you add "important" because you found it not to work without? Inline styles have a very high specificity, so it shouldn't be needed?

File: webapp/tests/forms/steps/test_lotse_steuerlotse_steps.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+
+from app.elster_client.elster_client import request_tax_offices
+from app.forms.steps.lotse.lotse_steuerlotse_steps import StepSteuernummer, LotseFormSteuerlotseStep
+
+
+class SummaryStep(object):
+    pass
+
+
+class TestLotseFormSteuerlotseStep:
```
Comment (nfelger) [inline webapp/tests/forms/steps/test_lotse_steuerlotse_steps.py:12]: OMG pytest tests in the wild 😱 😍

File: webapp/tests/forms/steps/test_lotse_steuerlotse_steps.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+    @pytest.fixture(autouse=True)
+    def attach_fixtures(self, app, test_request_context):
+        self.app = app
+        self.req = test_request_context
```
Comment (nfelger) [inline webapp/tests/forms/steps/test_lotse_steuerlotse_steps.py:16]: The `attach_fixtures` pattern is only necessary for adding pytest fixtures to unittest class-based test suites. I wouldn't recommend it for pure pytest tests.

Instead, specify on the test function itself which fixtures are used.

Three variants:

1) If you need the value of the fixture in the test, ["request" it by passing it as a parameter](https://docs.pytest.org/en/6.2.x/fixture.html#requesting-fixtures):
```py
def test_whatever(app):
    assert app.something() == 1
```

2) If you only need the side-effect **in a single test**, [use a decorator on the test](https://docs.pytest.org/en/6.2.x/fixture.html#use-fixtures-in-classes-and-modules-with-usefixtures):
```py
@pytest.mark.usefixtures("app")
def test_whatever():
    assert 2+2 = 5
```

3) If you only need the side-effect **for all tests a class**, decorate the class. Note that you can still request the fixture in individual tests if you need the value there.
```py
@pytest.mark.usefixtures("app")
class TestStuff:
    def test_whatever(self):
        assert 2+2 = 5
    def test_using_value(self, app):
        assert app.something() == 1
```

**Review by nfelger — CHANGES_REQUESTED** [review]
Big work in many areas! I love the added StepChooser, the pytest tests, the experiments with managing frontend state. I think the JS needs some changes (especially changing the click -> change handler), otherwise I think it's great.

There's a lot of JS logic that's covered only by a happy path acceptance test - that's the right trade-off at this point, but we should make sure the UI logic receives even thorougher-than-usual QA!

File: webapp/app/static/js/field_utils.js (hunk 17 above)
Comment (JulianRoesner) [reply webapp/app/static/js/field_utils.js:37]: Yes, I had to add it because it did not work otherwise. The problem is that some of the classes use important and therefore overwrites everything else.

File: webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py [hunk 20]
```diff
@@ -0,0 +1,149 @@
+from flask import request
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py:1]: Yes, I wanted to have one place for the new steps of the lotse flow (SteuerlotseSteps) to not mix them up with the old steps. But this probably was not the best way to organize that.

Either:
- We move the steps into the `personal_data_steps.py`
- We move this into `forms/steps/lotse/steuerlotse_steps/personal_data_steps`

File: webapp/tests/forms/steps/test_lotse_steuerlotse_steps.py (hunk 19 above)
Comment (JulianRoesner) [reply webapp/tests/forms/steps/test_lotse_steuerlotse_steps.py:16]: That makes so much more sense. Thank you for explaining that

File: webapp/app/templates/lotse/form_steuernummer.html (hunk 14 above)
Comment (JulianRoesner) [reply webapp/app/templates/lotse/form_steuernummer.html:63]: Yes, I overlooked that and I will change it to `let` and I agree with not using `var` anymore

File: webapp/tests/forms/flows/test_lotse_step_chooser.py (hunk 16 above)
Comment (JulianRoesner) [reply webapp/tests/forms/flows/test_lotse_step_chooser.py:18]: I agree

File: webapp/app/templates/lotse/form_steuernummer.html [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+            'TaxOfficeSelection': 'TaxOfficeSelection',
+            'TaxNumberInput': 'TaxNumberInput',
+            'RequestNewTaxNumber': 'RequestNewTaxNumber'
+        }
+        let taxNumberFormState = TAX_NUMBER_FORM_STATES.NothingSelected;
+
+        document.addEventListener('DOMContentLoaded', () => {
+            // Register event handlers
+            document.querySelectorAll('input[type=radio][name=steuernummer_exists]').forEach((field)=>{
+                field.addEventListener("click",() => {changeTaxNumberExists(); runThroughAllStates(); handleState();});
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/templates/lotse/form_steuernummer.html:47]: Frede mentioned something. I used change in the beginning. However, using `change` does not work here because we use bootstraps `data-toggle` for the yes-no-field. As far as I can tell, bootstrap overwrites and prevents the default behavior if data-toggle is not set on the target directly, which is the case for us (s. [here](https://github.com/minhur/bootstrap-toggle/issues/174#issuecomment-614745518) and [here](https://github.com/twbs/bootstrap-sass/blob/master/assets/javascripts/bootstrap/button.js#L115)). 
In the bootstrap [documentation](https://getbootstrap.com/docs/3.4/javascript/), they use Jquery `change` which works. 
As far as I can tell, we could either:
- get rid of bootstrap at this point
- change how the YesNo-Field is put together 
- we keep the click event

Personally, I would think that we should keep this in mind, when we compentonize this with React but keep the click event until then.

File: webapp/app/templates/lotse/form_steuernummer.html (hunk 4 above)
Comment (JulianRoesner) [reply webapp/app/templates/lotse/form_steuernummer.html]: See the answer below

File: webapp/app/templates/lotse/form_steuernummer.html (hunk 21 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/templates/lotse/form_steuernummer.html:47]: I think I disagree with you here. As Niko said, `click` won't work for users without a mouse, so e.g. screen reader users. And that they can use the site should in my opinion be more important than having a nice animation for sighted people. So I'd remove the `data-toggle` instead (or maybe even implement the animation ourselves?) You're very welcome to disagree with me again. :D

File: webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (nfelger) [reply webapp/app/forms/steps/lotse/lotse_steuerlotse_steps.py:1]: Gotcha - I was mainly talking about the name, but I agree that putting everything into one module isn't great. How about renaming the existing `forms/steps/lotse/` to `forms/steps/lotse_multi_step_flow_steps/` and putting the new steps into `forms/steps/lotse/personal_data_steps.py`? Would avoid my confusion.

But moving the steps into `personal_data_steps.py` would also work.

File: webapp/app/templates/lotse/form_steuernummer.html (hunk 21 above)
[the remark this answers is not in the document]
Comment (nfelger) [reply webapp/app/templates/lotse/form_steuernummer.html:47]: I agree with @frederike-ramin that using `click` is out of the question.

How about just using jQuery for that one case with a descriptive comment? We'll have to go over this again anyway once we reactify it and then we can decide wholesale whether we want to drop bootstrap's JS components or use them via something like [react-bootstrap](https://react-bootstrap.github.io/) (which reimplements bootstrap's JS without jquery).

[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/templates/lotse/form_steuernummer.html:47]: > And that they can use the site should in my opinion be more important than having a nice animation for sighted people. 

100% agree. I forgot about that aspect while diving into the reason of the problem.

> How about just using jQuery for that one event listener (with a descriptive comment)? We'll have to go over this again anyway once we reactify it and then we can decide wholesale whether we want to drop bootstrap's JS components or use them via something like react-bootstrap (which reimplements bootstrap's JS without jquery).

I would prefer that approach compared to removing the `data-toggle` because this would add a whole lot of extra javascript.

**Review by nfelger — APPROVED** [review]
